README.md
# MirHamit/PodBankService

[![Latest Version on Packagist](https://img.shields.io/packagist/v/vendor_slug/package_slug.svg?style=flat-square)](https://packagist.org/packages/mirhamit/pod-bank-service)
[![Total Downloads](https://img.shields.io/packagist/dt/vendor_slug/package_slug.svg?style=flat-square)](https://packagist.org/packages/mirhamit/pod-bank-service)

---
این پکیج برای استفاده از سرویس بانکی پاد در لاراول تهیه شده است، شما میتوانید مستندات کامل را از وبسایت پادیوم دریافت
کنید

استفاده از این پکیج بلامانع است، اما به دلیل حساسیت بالای تراکنش ها و انتقال وجه، منتشر کننده این پکیج هیچ گونه مسئولیتی در قبال مسائل مربوطه ندارد.


[راهنمای کلی سرویس](https://podium.ir/services/3292395/%D8%B3%D8%B1%D9%88%DB%8C%D8%B3-%D9%87%D8%A7%DB%8C-%D8%A8%D8%A7%D9%86%DA%A9%DB%8C/document)

## نصب

نصب لاراول

```bash
laravel new laravel-pod-bank-service-package
cd laravel-pod-bank-service-package
```

نصب پکیج

```bash
composer require mirhamit/pod-bank-service
```

## راه اندازی و تنظیمات

استخراج فایل کانفیگ و اضافه کردن رمز عبور و کلید دسترسی و دیگر موارد مورد نیاز پکیج و وب سرویس پادیوم در فایل کانفیگ

```bash
php artisan vendor:publish --tag=pod-config
```

و اگه به هر دلیلی دوست نداشتین رمز عبور و ... رو توی فایل کانفیگ بزارین میتونین به صورت دستی کلید های مورد نیاز رو در
فایل

.env

موجود بزارین و موارد مورد نیاز رو جایگزین کنین

```bash
pod_UserName="نام کاربری که از بانک دریافت کرده‌اید"
pod_DepositNumber="شماره حسابی که به بانک معرفی کرده‌اید"
pod_Sheba="شماره شبای حسابی که به بانک معرفی کرده‌اید"
pod_token="کلیدی که از قسمت ورود یکپارچه لیست کلیدها یا کلید جدید دریافت میکنید"

# کلیدهایی که طبق راهنمای موجود در بالای همین صفحه و وبسایت پادیوم دریافت میکنین

pod_scApiKeyEstelamSheba="سپرده / استعلام - اطلاعات شبا"
pod_scApiKeyMojudi="سپرده - موجودی حساب"
pod_scApiKeyEnteghalPaya="سپرده - انتقال وجه - ( داخلی و پایا )"
pod_scApiKeySuratHesab="سرویس صورتحساب و وضعیت انتقال"

pod_scProductIdEstelamSheba=1115396
pod_scProductIdMojudi=1077449
pod_scProductIdEnteghalPaya=1076566
pod_scProductIdSuratHesab=1077467
```

---

## استفاده

[راهنمایی](https://podium.ir/services/3292395/%D8%B3%D8%B1%D9%88%DB%8C%D8%B3-%D9%87%D8%A7%DB%8C-%D8%A8%D8%A7%D9%86%DA%A9%DB%8C/document)


###سرویس دریافت اطلاعات شماره شبا (همه ی بانکها)
شناسه سرویس : 1115396

```php
use MirHamit\PodBankService\Services\SavingAccount;

$account = new SavingAccount();
return $account->estelamSheba($sheba, $paymentId);
```


###سرویس دریافت موجودی حساب
شناسه سرویس : 1077449

```php
use MirHamit\PodBankService\Services\SavingAccount;

$account = new SavingAccount();
return $account->mojudi()
```


###سرویس دریافت صورتحساب حساب
شناسه سرویس : 1077467

```php
use MirHamit\PodBankService\Services\SavingAccount;

$account = new SavingAccount();
return $account->suratHesab($fromDate, $toDate, $fromHour, $toHour, $resultCount);
```

###سرویس انتقال وجه پایا بین بانکی ( با قابلیت انجام انتقال وجه داخلی بانک پاسارگاد )
شناسه سرویس : 1076566

```php
use MirHamit\PodBankService\Services\SavingAccount;

$account = new SavingAccount();

return $account->enteghalVajhPaya(
        1000,
        'IR200610000000700813563379',
        'نام و نام خانوادگی',
        '1400/10/11', //تاریخ
        'شرح انتقال',
        'شرح مقصد',
        '123123123123', //شناسه‌ی واریز
        '' //شناسه یکتای تراکنش
    );
```

## خروجی توابع

نمونه خروجی یکی از توابع در زیر آمده است که در صورت نیاز به راهنمایی بیشتر میتوانید به راهنمای ذکر شده در بالای این صفحه
رجوع کنید


---

## Tests

```text
As Soon As Possible
```

---

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

---

## Security Vulnerabilities

Please review and check security vulnerabilities and report them in issues section.

---

## Credits

- [Həmid Musəvi](https://github.com/mirhamit)
- [All Contributors](../../contributors)

---

## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.


CHANGELOG.md
# Changelog
این پکیج برای استفاده از سرویس بانکی پاد در لاراول تهیه شده است، شما میتوانید مستندات کامل را از وبسایت پادیوم دریافت
کنید

[راهنمای کلی سرویس](https://podium.ir/services/2974845/%D8%B3%D8%B1%D9%88%DB%8C%D8%B3%E2%80%8C%D9%87%D8%A7%DB%8C%20%D8%A8%D8%A7%D9%86%DA%A9%20%D9%BE%D8%A7%D8%B3%D8%A7%D8%B1%DA%AF%D8%A7%D8%AF)

## 1.1.2 - 2022-01-02
- ارسال کد بانک به صورت خودکار از شماره شبا

## 1.1.1 - 2022-01-01
- سازگاری با مستندات و ای پی آی های جدید سرویس های بانکی پادیوم


## 1.0.2 - 2021-11-20
- رفع مشکل استعلام انتقال وجه


## 1.0.1 - 2021-11-20
- رفع مشکل نصب اولیه


## 1.0.0 - 2021-11-18
- اولین ورژن


src/Services/DataSender.php
<?php
/**
 * @author Həmid Musəvi <w1w@yahoo.com>
 * 11/16/21
 */

namespace MirHamit\PodBankService\Services;

use Illuminate\Support\Facades\Http;

class DataSender
{
    private $requestUrl = 'https://api.pod.ir/srv/sc/nzh/doServiceCall';

    public static function sendRequest($headers, $requestData, bool $bankData = true)
    {
        return (new DataSender)->sendHttpRequest($headers, $requestData, $bankData);
    }

    public static function sign($dataToSign): string
    {
        return (new DataSender)->signData($dataToSign);
    }

    private function sendHttpRequest($headers, $requestData)
    {
        $req = Http::asForm()->withHeaders($headers)->post($this->requestUrl, $requestData);
        $result = $req->body();
        if ($this->isJson($req->body())) {
            $result = json_decode($req->body());
            if (!$result->hasError) {
                if ($this->isJson($result->result->result)) {
                    $result->result->result = json_decode($result->result->result);
                }
            }
        }
        return $result;
    }

    private function signData($dataToSign): string
    {
        $signature = null;

        openssl_sign($dataToSign, $signature, config('pod.pod_PRIVATE_KEY'), OPENSSL_ALGO_SHA1);

        return base64_encode($signature);
    }

    private function isJson($string)
    {
        json_decode($string);
        return json_last_error() === JSON_ERROR_NONE;
    }
}


src/Services/SavingAccount.php
<?php

namespace MirHamit\PodBankService\Services;

use Carbon\Carbon;
use Exception;
use Illuminate\Support\Str;

/**
 * @author Həmid Musəvi <w1w@yahoo.com>
 * 11/14/21
 */
class SavingAccount
{

    protected $dataSender;


    /**
     * Shenase Service 1077449
     * /mojudi
     * */
    public function mojudi()
    {
        if (!config('pod.pod_token') || !config('pod.pod_Services.scProductIdMojudi') || !config('pod.pod_Services.scApiKeyMojudi')) {
            throw new Exception('لطفا مقادیر مورد نیاز سرویس را بررسی کنید', 503);
        }
        $data = "{".'"DepositNumber":"'.config('pod.pod_DepositNumber').'"'."}";


        $headers = [
            '_token_'        => config('pod.pod_token'),
            '_token_issuer_' => '1',
            'Content-Type'   => 'application/x-www-form-urlencoded',
        ];
        $requestData = [
            'scProductId' => config('pod.pod_Services.scProductIdMojudi'),
            'scApiKey'    => config('pod.pod_Services.scApiKeyMojudi'),
            'request'     => $data,
        ];
        return DataSender::sendRequest($headers, $requestData);
    }


    /**
     * Shenase Service 1115396
     * /estelam/IR111111111111111111111111
     */
    public function estelamSheba($sheba, $paymentId = null)
    {

        if (!config('pod.pod_token') || !config('pod.pod_Services.scProductIdEstelamSheba') || !config('pod.pod_Services.scApiKeyEstelamSheba')) {
            throw new Exception('لطفا مقادیر مورد نیاز سرویس را بررسی کنید', 503);
        }
        $data = "{".
            '"Iban":"'.$sheba.'",'.
            '"PaymentId":"'.$paymentId.'"'
            ."}";
        $headers = [
            '_token_'        => config('pod.pod_token'),
            '_token_issuer_' => '1',
            'Content-Type'   => 'application/x-www-form-urlencoded',
        ];
        $requestData = [
            'scProductId' => config('pod.pod_Services.scProductIdEstelamSheba'),
            'scApiKey'    => config('pod.pod_Services.scApiKeyEstelamSheba'),
            'request'     => $data,
        ];

        return DataSender::sendRequest($headers, $requestData);
    }


    /**
     * Shenase Service 1076566
     * /enteghalVajhPaya
     */
    public function enteghalVajhPaya(
        $amount,
        $destinationIban,
        $recieverFullName,
        $transactionDate,
        $description,
        $srcComment,
        $transactionBillNumber = null,
        $transactionId = null
    ) {
        if (!config('pod.pod_token') || !config('pod.pod_Services.scProductIdEnteghalPaya') || !config('pod.pod_Services.scApiKeyEnteghalPaya')) {
            throw new Exception('لطفا مقادیر مورد نیاز سرویس را بررسی کنید', 503);
        }
        if (!$transactionId) {
            $orgCode = "4321";
            $randomString = Str::random(8);
            $dateTime = Carbon::now()->getPreciseTimestamp(3);
            $asci = 0;
            foreach (str_split($orgCode) as $item) {
                $asci += ord($item);
            }
            foreach (str_split($randomString) as $item) {
                $asci += ord($item);
            }
            foreach (str_split($dateTime) as $item) {
                $asci += ord($item);
            }
            $transactionId = "{$orgCode}-{$randomString}-{$dateTime}-{$asci}";
        }
        $data = "{".
            '"Amount":"'.$amount.'",'. //مبلغ تراکنش
            '"SourceDepNum":"'.config('pod.pod_DepositNumber').'",'. //شماره سپرده ی مبدا
            '"DestinationIban":"'.$destinationIban.'",'. //شماره شبای مقصد
            '"RecieverFullName":"'.$recieverFullName.'",'. //نام گیرنده‌ی وجه
            '"TransactionDate":"'.$transactionDate.'",'.
            '"Description":"'.$description.'",'. //شرح انتقال
            '"TransactionId":"'.$transactionId.'",'. //شناسه یکتای تراکنش
            '"IsAutoVerify":"true",'. //تایید یا عدم تایید خودکار
            '"SrcComment":"'.$srcComment.'",'. //شرح مبدا
            '"SenderReturnDepositNumber":"'.config('pod.pod_DepositNumber').'",'. //شماره سپرده‌ی بازگشت وجه
            '"CustomerNumber":"'.config('pod.pod_customerNumber').'",'. //شماره مشتری
            '"DestBankCode":"'.substr($destinationIban, 5, 2).'",'. //کد بانک مقصد
            '"TransactionBillNumber":"'.$transactionBillNumber.'"' //شناسه‌ی واریز
            ."}";
        $headers = [
            '_token_'        => config('pod.pod_token'),
            '_token_issuer_' => '1',
            'Content-Type'   => 'application/x-www-form-urlencoded',
        ];
        $requestData = [
            'scProductId' => config('pod.pod_Services.scProductIdEnteghalPaya'),
            'scApiKey'    => config('pod.pod_Services.scApiKeyEnteghalPaya'),
            'request'     => $data,
        ];
        return DataSender::sendRequest($headers, $requestData);
    }


    /**
     * Shenase Service 1077467
     * /suratHesab/1400-10-08/1400-10-08/00-00-00/23-59-59
     */
    public function suratHesab(
        $fromDate,
        $toDate,
        $fromHour = "00:00:00",
        $toHour = "23:59:59",
        $resultCount = 99999999
    ) {
        if (!config('pod.pod_token') || !config('pod.pod_Services.scProductIdSuratHesab') || !config('pod.pod_Services.scApiKeySuratHesab')) {
            throw new Exception('لطفا مقادیر مورد نیاز سرویس را بررسی کنید', 503);
        }
        if (Str::contains($fromDate, '-')) {
            $fromDate = Str::replace('-', '/', $fromDate);
        }
        if (Str::contains($toDate, '-')) {
            $toDate = Str::replace('-', '/', $toDate);
        }
        if (Str::contains($fromHour, '-')) {
            $fromHour = Str::replace('-', ':', $fromHour);
        }
        if (Str::contains($toHour, '-')) {
            $toHour = Str::replace('-', ':', $toHour);
        }
        $data = "{".
            '"DepositNumber":"'.config('pod.pod_DepositNumber').'",'.
            '"FromDate":"'.$fromDate.'",'.
            '"ToDate":"'.$toDate.'",'.
            '"FromTime":"'.$fromHour.'",'.
            '"ToTime":"'.$toHour.'",'.
            '"ResultCount":"'.$resultCount.'"'
            ."}";
        $headers = [
            '_token_'        => config('pod.pod_token'),
            '_token_issuer_' => '1',
            'Content-Type'   => 'application/x-www-form-urlencoded',
        ];
        $requestData = [
            'scProductId' => config('pod.pod_Services.scProductIdSuratHesab'),
            'scApiKey'    => config('pod.pod_Services.scApiKeySuratHesab'),
            'request'     => $data,
        ];
        //        return $requestData;
        return DataSender::sendRequest($headers, $requestData);
    }


}


config/pod.php
<?php

/**
 * @author Həmid Musəvi <w1w@yahoo.com>
 * 11/14/21
 */

return [
    "pod_UserName"       => env('pod_UserName', ''),
    "pod_DepositNumber"  => env('pod_DepositNumber', ''),
    "pod_customerNumber" => env('pod_customerNumber', ''),
    "pod_Sheba"          => env('pod_Sheba', ''),
    "pod_token"          => env('pod_token', ''),
    "pod_PRIVATE_KEY"   => env('pod_PRIVATE_KEY', ''),

    'pod_Services' => [
        'scProductIdEstelamSheba' => env('pod_scProductIdEstelamSheba', '1115396'),
        'scApiKeyEstelamSheba'    => env('pod_scApiKeyEstelamSheba', ''),

        'scProductIdMojudi' => env('pod_scProductIdMojudi', '1077449'),
        'scApiKeyMojudi'    => env('pod_scApiKeyMojudi', ''),

        'scProductIdEnteghalPaya' => env('pod_scProductIdEnteghalPaya', '1076566'),
        'scApiKeyEnteghalPaya'    => env('pod_scApiKeyEnteghalPaya', ''),

        'scProductIdSuratHesab' => env('pod_scProductIdSuratHesab', '1077467'),
        'scApiKeySuratHesab'    => env('pod_scApiKeySuratHesab', ''),

    ]
];
